java/code/src/com/suse/manager/webui/services/OidcAuthHandler.java

```diff
@@ -0,0 +1,287 @@
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package com.suse.manager.webui.services;
+
+import com.redhat.rhn.common.conf.ConfigDefaults;
+import com.redhat.rhn.common.conf.ConfigException;
+import com.redhat.rhn.common.util.http.HttpClientAdapter;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.util.EntityUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jose4j.jwa.AlgorithmConstraints;
+import org.jose4j.jwk.HttpsJwks;
+import org.jose4j.jws.AlgorithmIdentifiers;
+import org.jose4j.jwt.JwtClaims;
+import org.jose4j.jwt.MalformedClaimException;
+import org.jose4j.jwt.consumer.InvalidJwtException;
+import org.jose4j.jwt.consumer.JwtConsumer;
+import org.jose4j.jwt.consumer.JwtConsumerBuilder;
+import org.jose4j.keys.resolvers.HttpsJwksVerificationKeyResolver;
+import org.jose4j.keys.resolvers.VerificationKeyResolver;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+/**
+ * Handles OpenID Connect (OIDC) authentication.
+ */
+public class OidcAuthHandler {
+
+    public static final String OIDC_DISCOVERY_PATH = "/.well-known/openid-configuration";
+
+    private static final Logger LOG = LogManager.getLogger(OidcAuthHandler.class);
+
+    private boolean oidcEnabled;
+    private String issuer;
+    private String jwksUri;
+    private String audience;
+    private String usernameClaim;
+
+    private JwtConsumer jwtConsumer;
+    private HttpClientAdapter httpClient;
+
+    /**
+     * Constructs an OidcAuthHandler and loads configuration.
+     */
+    public OidcAuthHandler() {
+        this(null, new HttpClientAdapter());
+    }
+
+    /**
+     * Constructs an OidcAuthHandler with a custom {@link VerificationKeyResolver}.
+     *
+     * @param keyResolver the verification key resolver, if {@code null} a default one is created
+     * @param httpClientIn the HTTP client to use when fetching from the discovery endpoint
+     */
+    public OidcAuthHandler(VerificationKeyResolver keyResolver, HttpClientAdapter httpClientIn) {
+        httpClient = httpClientIn;
+
+        try {
+            loadConfiguration();
+            if (!isOidcEnabled()) {
+                return;
+            }
+        }
+        catch (URISyntaxException eIn) {
+            throw new ConfigException("Malformed URI in the OIDC configuration.");
+        }
+        VerificationKeyResolver resolver = keyResolver;
+        if (resolver == null) {
+            HttpsJwks httpsJwks = new HttpsJwks(jwksUri);
```

Comment:
Not sure if this is the case, but you probably can cache this public key as these are infrequently rotated.